### PR TITLE
Dao Interface

### LINKED ISSUE #542: Proposal: using interfaces for DAO operations to allow simple unit tests
> # Feature:
When looking at how I would implement the last_operation description https://github.com/openshift/ansible-service-broker/pull/537
I wondered how I would go about adding tests for it also. I know there are a number of end to end tests that are used, but wondered what appetite there was to improving the ability to add unit tests?

If we were to use interfaces to pick out the relevant parts of the DAO object, I think it would give a good starting place.

I have created a reference PR to show what I mean and am interested in thoughts and feedback.

### PR DESCRIPTION
**Pull request by shawn-hurley**
**Describe what this PR does and why we need it**:
Having an interface for the DAO will allow us to switch persistent layers. This will also allow for unit testing in other parts of the code base.

Changes proposed in this pull request
 - move etcd dao into dao/etcd package
 - everything should now use the DAO interface

**Does this PR depend on another PR (Use this to track when PRs should be merged)**
depends-on <PR>
fix for PR #722 Work Item 1
**Which issue this PR fixes (This will close that issue when PR gets merged)**
This fixes #542

### CODE REVIEW

**Comment by shawn-hurley** [comment]
@maleck13 Sorry I am not fully following your recommendation regarding interfaces. 

What I am seeing is `dao` is the consumer of `dao/etcd`(the provider) and therefore defines the interface. 

Is that different then what you are referring to?

**Comment by maleck13** [comment]
Sorry @shawn-hurley  I will give a more concrete  example of what I mean.

The broker package depends on the DAO. So in my suggestion the broker package would now define the interface that it needs as it knows its requirements best and so would no longer depend on
``` "github.com/openshift/ansible-service-broker/pkg/dao" ```
as the interface is defined within its own package. Does that make what I mean clearer. As mentioned this is only a suggestion

**Comment by shawn-hurley** [comment]
Ah ok, that does make more sense now thank you.

I would like to hear more about that helps prevent circular dependencies. 
Also, is their an example in the standard library of using the interface like this? I am only asking because that might help me match use cases?

File: pkg/dao/dao.go [hunk 1]
```diff
@@ -17,407 +17,83 @@
...
-	for i, payload := range *payloads {
-		spec := &apb.Spec{}
-		apb.LoadJSON(payload, spec)
-		specs[i] = spec
-		log.Debug("Batch idx [ %d ] -> [ %s ]", i, spec.ID)
-	}
-
-	return specs, nil
+func NewDao() (Dao, error) {
+	return etcd.NewDao()
```
Comment (jmrodri) [inline pkg/dao/dao.go]: Is the idea we'll have multiples from here?

Comment (shawn-hurley) [reply pkg/dao/dao.go]: Yep, that was the initial plan/thought process

**Comment by jmrodri** [comment]
@eriknelson @shawn-hurley as far as names, the only thing that is descriptive is DAL since it is a data access layer vs DAO which usually means data access object.

**Comment by jmrodri** [comment]
@eriknelson @shawn-hurley correction I do have a name suggestion: AXON

http://www.dictionary.com/browse/axon

> the long threadlike part of a nerve cell along which impulses are conducted from the cell body to other cells.

**Comment by jmrodri** [comment]
Are we holding off because of naming? or are we good?

**Comment by maleck13** [comment]
@shawn-hurley hope this is an interesting discussion ( I think so) but not my intention to hijack the PR :)
**TLDR**
- It is not a hard black and white rule. 
- You could refactor a package consuming an interface to declare its own version easily if it became and issue
- It is really just a pattern I have found useful and to work well

**Standard Lib:**
I haven't gone hunting through the standard library to check, however i know the ```io``` package declares the ```Writer``` and ```Reader```  interface and consumes them heavily throughout that package. 
So in my opinion, the io package knows what it wants in order to consume a reader or a writer, so it is the package that declares the interfaces. However in this case, it is an exported interface as it is generic enough that it can be useful outside of the io package. If it was not useful outside of the io package I don't think it would be exported and would just be an internal interface. There are some of these in the io package also.

On the circular dependencies: (this really is less of a thing but a side benefit)
By declaring the interface within the consuming package you reduce the the chance of a circular dependency as you have removed an import of an external package.
